ee/clickhouse/timer.py

```diff
@@ -0,0 +1,105 @@
+import logging
+import uuid
+from collections import OrderedDict
+from functools import partial
+from threading import Condition, Thread
+from time import time
+from typing import Callable, Dict, Optional, Tuple
+
+from django.conf import settings
+
+logger = logging.getLogger(__name__)
+
+
+class TimerTask:
+    id: str
+    done: bool
+
+    def __init__(self, callback: Callable, *args, **kwargs):
+        self.callback = partial(callback, *args, **kwargs)
+        self.id = str(uuid.uuid4())
+        self.done = False
+
+    def run(self):
+        self.done = True
+        try:
+            self.callback()
+        except Exception as err:
+            logger.warn("TimerTask failed, ignoring error", err)
+
+
+class SingleThreadedTimer(Thread):
+    def __init__(self, timeout_ms: int, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.setDaemon(True)
+
+        self.timeout_ms = timeout_ms
+        self.started = False
+        self.lock = Condition()
+        self.tasks: OrderedDict = OrderedDict()
+
+    def schedule(self, callback: Callable, *args, **kwargs) -> TimerTask:
+        self.start()
+
+        with self.lock:
+            task = TimerTask(callback, *args, **kwargs)
+            self.tasks[task.id] = (task, time())
+            self.lock.notify()
+
+            return task
+
+    def cancel(self, task: TimerTask) -> None:
+        with self.lock:
+            del self.tasks[task.id]
+            self.lock.notify()
+
+    def start(self):
+        if not self.started:
+            self.started = True
+            super().start()
+
+    def run(self):
+        while True:
+            job = None
+            with self.lock:
+                sleep = self._sleep_time()
+                if len(self.tasks) == 0:
+                    self.lock.wait()
+                elif sleep > 0:
+                    self.lock.wait(sleep)
+                else:
+                    _, (job, _) = self.tasks.popitem(last=False)
+
+            if job is not None:
+                job.run()
+
+    def _next_task(self) -> Optional[Tuple[TimerTask, float]]:
+        for _, task_and_time in self.tasks.items():
+            return task_and_time
+        return None
+
+    def _sleep_time(self) -> float:
+        next_task = self._next_task()
+        if next_task is None:
+            return 0
+        else:
+            _, start_time = next_task
+            return start_time + self.timeout_ms / 1000.0 - time()
+
+
+class TestSingleThreadedTimer(SingleThreadedTimer):
```

Comment:
Is there a reason we've decided not to test this functionality?

Reply:
You mean why a different timer in tests or why not tests for the class?

For the first, it would add load to an already loaded system
For the second, this is auxilary enough and isolated not to affect anything, plus testing sleep and thread-related things would probably have convoluted the code a bit.

Reply:
First makes sense!

For second: An option is to just separately test the `SingleThreadedTimer` class, to ensure it works as we expect.
(Something like here: https://github.com/PostHog/posthog/blob/master/ee/clickhouse/test/test_client.py )

While I agree that it's auxillary enough to not affect anything else, I also want to test for its own correctness, as it would imply wrong metrics if we get this wrong 😅 .

Reply:
Yeah, second is me being lazy. A justification I tried express earlier is that timing-related testing is tricky - very very easy to introduce a new flaky test due to one system being faster than the other and then slow down the whole suite due to excessive `sleep`s. 

The "proper" way is by mocking timers and `sleep` calls but that often ends up convoluting the application code. In js you have https://jestjs.io/docs/timer-mocks but not aware of anything built-in to unittest.

Reply:
I wrote a few that seem reasonable (enough time between events). Mind if I push here?

Reply:
Feel free to remove if you're not ok with these, just pushed^

Reply:
Yeah that test is exactly what I meant by it's prone to flake.